storage/src/app_hash.rs

```diff
@@ -0,0 +1,135 @@
+use ibc::core::ics23_commitment::{commitment::CommitmentPrefix, specs::ProofSpecs};
+use jmt::{storage::TreeReader, RootHash};
+use once_cell::sync::Lazy;
+use penumbra_proto::Message;
+use sha2::{Digest, Sha256};
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct AppHash(pub [u8; 32]);
+
+// the app hash of penumbra's state is defined as SHA256("penumbra-root" || jmt.root_hash())
+impl From<RootHash> for AppHash {
+    fn from(r: RootHash) -> Self {
+        let mut h = Sha256::new();
+        h.update(APPHASH_DOMSEP);
+        h.update(r.0);
+
+        AppHash(h.finalize().into())
+    }
+}
+
+fn apphash_spec() -> ics23::ProofSpec {
+    ics23::ProofSpec {
+        leaf_spec: Some(ics23::LeafOp {
+            prefix: vec![],
+            hash: ics23::HashOp::Sha256.into(),
+            length: ics23::LengthOp::NoPrefix.into(),
+            prehash_key: ics23::HashOp::NoHash.into(),
+            prehash_value: ics23::HashOp::NoHash.into(),
+        }),
+        inner_spec: Some(ics23::InnerSpec {
+            hash: ics23::HashOp::NoHash.into(),
+            child_order: vec![0, 1],
+            child_size: 32,
+            empty_child: vec![],
+            min_prefix_length: 0,
+            max_prefix_length: 0,
+        }),
+        min_depth: 0,
+        max_depth: 2,
+    }
+}
+
+static APPHASH_DOMSEP: &str = "penumbra_root";
```

Comment:
not sure if this is the best label, it doesn't matter but would be cool to have something cool here :D